centinel/client.py

```diff
@@ -98,39 +90,81 @@ def run(self, data_dir=None):
                          "%s" % (self.config['dirs']['results_dir']))
             os.makedirs(self.config['dirs']['results_dir'])
 
-        experiments = self.load_experiments()
-        experiments_subset = experiments.items()
+        experiments_set = self.experiments.items()
 
-        if self.config['experiments']['random_subsetting'] and \
-               self.config['experiments']['random_subset_size'] < \
-                   len(experiments.items()):
-            experiments_subset = [
-                experiments.items()[i] for i in sorted(
-                    random.sample(xrange(len(experiments.items())),
-                                  self.config['experiments']
-                                             ['random_subset_size']))
-            ]
+        # load scheduler information
+        sched_filename = os.path.join(self.config['dirs']['experiments_dir'],
+                                      'scheduler.info')
+        sched_info = {}
+        if os.path.exists(sched_filename):
+            with open(sched_filename, 'r') as file_p:
+                sched_info = json.load(file_p)
 
-        for name, Exp in experiments_subset:
+        for name in sched_info:
 
+            # check if we should preempt on the experiment (if the
+            # time to run next is greater than the current time) and
+            # store the last run time as now
+            #
+            # Note: if the experiment is not in the scheduler, then it
+            # will not be run at all.
+            run_next = sched_info[name]['last_run']
+            run_next += sched_info[name]['frequency']
+            if run_next > time.time():
+                continue
+
+            # backward compatibility with older-style scheduler
+            if 'python_exps' not in sched_info[name]:
+                self.run_exp(name)
+            else:
+                for python_exp, exp_config in sched_info[name]['python_exps'].items():
+                    self.run_exp(python_exp, exp_config)
+
+            sched_info[name]['last_run'] = time.time()
+
+        # write out the updated last run times
+        with open(sched_filename, 'w') as file_p:
+            json.dump(sched_info, file_p, indent = 2,
+                      separators=(',', ': '))
+
+        self.consolidate_results()
+
+        logging.info("Finished running experiments. Look in %s for "
+                     "results." % (self.config['dirs']['results_dir']))
+
+    def run_exp(self, name, exp_config=None):
+        if name not in self.experiments:
+            logging.error("Experiment file %s not found! Skipping." % (name))
+        else:
+            Exp = self.experiments[name]
+        #for name, Exp in experiments_subset:
```

Comment:
can adjust the indent level for this comment please (or remove it because this doesn't look like it is supposed to be here)